proposals/030-coherent-pointers.md

```diff
@@ -0,0 +1,203 @@
+	  
+`MakePointerAvailable` is for memory stores of non textures, `OpStore`, `OpCooperativeMatrixStoreKHR` and `OpCooperativeVectorStoreNV`.. 
+
+`MakePointerVisible` is for memory loads of non textures, `OpLoad`, `OpCooperativeMatrixLoadKHR` and `OpCooperativeVectorLoadNV`.
+
+ `MakeTexelAvailableKHR` is for memory stores of textures, `OpImageWrite`, `OpImageSparseLoad`.
+
+ `MakeTexelVisibleKHR` is for memory loads of textures, `OpImageRead`, `OpImageSparseRead`.
+
+Additionally,  `MakePointer{Visible,Available}` support usage in `OpCopyMemory` and `OpCopyMemorySized`.
+
+### Example
+
+The simple use-case of this feature can be modeled with the following example: (1) We have **thread1** and **thread2** both reading/writing to the same `RWStructuredBuffer`. (2) **thread1** `OpStore`’s non-coherently into the buffer. (3) if **thread2** uses an `OpLoad` on the texture they may not see the change **thread1** made for 2 reasons:
+
+1) **thread1** does not promise its writes are visible to other threads. Cached writes may not immediately flush to device memory.  
+2) **thread2** may load from a cache, not device memory. This means we will not see the new value because the new value was written to device memory, not the intermediate cache.
+
+If we specify `MakePointerAvailable/MakePointerVisible` with `OpStore`/`OpLoad`  to the memory scope `QueueFamily` we will solve this problem since we are flushing changes to a memory scope shared by the two threads. This additionally may be faster than the alternative of flushing to `Device` memory since a `QueueFamily` is a tighter scope than `Device`.
+
+## Proposed Solution
+
+### Frontend For Coherent Pointers
+
+We propose to implement coherence on a per-operation level for only SPIR-V targets. This will be accomplished through modifying `Ptr` to include the new generic argument `CoherentScope coherentScope`.
+
+```c#
+public enum CoherentScope
+{
+    NotCoherent = 0xFF,
+    CrossDevice = MemoryScope::CrossDevice,
+    Device = MemoryScope::Device,
+    Workgroup = MemoryScope::Workgroup,
+    Subgroup = MemoryScope::Subgroup,
+    Invocation = MemoryScope::Invocation,
+    QueueFamily = MemoryScope::QueueFamily,
+    ShaderCallKHR = MemoryScope::ShaderCallKHR,
+    //...
+}
+
+__generic<T, uint64_t addrSpace=AddressSpace::UserPointer, CoherentScope coherentScope=CoherentScope.NotCoherent>
+struct Ptr
+{
+    ...
+}
+```
+
+If `coherentScope` is not `CoherentScope.NotCoherent`, all accesses to memory through this pointer will be considered coherent to the specified memory scope (example: `CoherentScope.Device` is coherent to the memory scope of `Device`).
+
+We will also provide a type alias for user-convenience.
+
+```c#
+__generic<T, CoherentScope coherentScope>
+typealias CoherentPtr = Ptr<T, AddressSpace::UserPointer, coherentScope>;
```

Comment:
I think this should be:
```
typealias CoherentPtr<T> = Ptr<T, AddressSpace::UserPointer, MemoryScope::CrossDevice>;
```

For anything else, the user should just use `Ptr<>` type.